hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/metrics/UserMetricsContext.java

```diff
@@ -0,0 +1,208 @@
+        if (res == null) {
+            res = initMetric(name, unit);
+            if (onlyName == null) {
+                // the first and so far the only metric
+                onlyName = name;
+                onlyMetric = res;
+            } else {
+                // 2 or more metrics
+                if (metrics == null) {
+                    metrics = new HashMap<>();
+                    metrics.put(onlyName, onlyMetric);
+                    onlyName = null;
+                    onlyMetric = null;
+                }
+                metrics.put(name, res);
+            }
+        }
+        return res;
+    }
+
+    private Metric initMetric(String name, Unit unit) {
+        if (probeBuilder == null) {
+            // used when metrics are disabled
+            return new NonRegisteredMetric(name);
+        } else {
+            RegisteredMetric metric = new RegisteredMetric(name);
+            LongProbeFunction<ProcessorTasklet> longProbeFunction = t -> metric.get();
+            probeBuilder.register(source, name, level, toProbeUnit(unit), longProbeFunction);
+            return metric;
+        }
+    }
+
+    private ProbeUnit toProbeUnit(Unit unit) {
+        switch (unit) {
+            case COUNT:
+                return ProbeUnit.COUNT;
+            case MS:
+                return ProbeUnit.MS;
+            case BYTES:
+                return ProbeUnit.BYTES;
+            default:
+                throw new RuntimeException("Unhandled metrics unit " + unit);
+        }
+    }
+
+    private interface Metric extends Counter, Gauge {
+
+        long get();
+
+    }
+
+    private static final class RegisteredMetric implements Metric {
+
+        private static final AtomicLongFieldUpdater<RegisteredMetric> VOLATILE_VALUE_UPDATER =
+                AtomicLongFieldUpdater.newUpdater(RegisteredMetric.class, "value");
+
+        private final String name;
+        private volatile long value;
+
+        RegisteredMetric(String name) {
+            this.name = name;
+        }
+
+        @Nonnull @Override
+        public String name() {
+            return name;
+        }
+
+        @Override
+        public void set(long newValue) {
+            VOLATILE_VALUE_UPDATER.lazySet(this, newValue);
+        }
+
+        @Override
+        public void inc() {
+            VOLATILE_VALUE_UPDATER.lazySet(this, value + 1);
+        }
+
+        @Override
+        public void add(long increment) {
+            VOLATILE_VALUE_UPDATER.lazySet(this, value + increment);
+        }
+
+        @Override
+        public void dec() {
+            VOLATILE_VALUE_UPDATER.lazySet(this, value - 1);
+        }
+
+        @Override
+        public void sub(long decrement) {
+            VOLATILE_VALUE_UPDATER.lazySet(this, value - decrement);
+        }
+
+        @Override
+        public long get() {
+            return value;
+        }
+    }
+
+    private static final class NonRegisteredMetric implements Metric {
```

Comment:
why do you need this? It would be simpler just to use the same object even if metrics are disabled. Seems like unnecessary over-optimization.